File: utils/execute.go
package utils

import (
	"fmt"
	flag "github.com/spf13/pflag"
	"io"
	"log"
	"path/filepath"
	"runtime"
	"time"
)

func GetFlags() (input string, part string) {
	flag.StringVar(&input, "input", "", "input file path")
	flag.StringVar(&part, "part", "one", "run part [one|two]")

	var verbose bool
	flag.BoolVarP(&verbose, "verbose", "v", false, "whether to show debug output")

	flag.Parse()

	if !verbose {
		log.SetOutput(io.Discard)
	}

	return
}

func Execute(fn func(Input) interface{}, input string) (result interface{}, elapsed time.Duration) {
	start := time.Now()
	result = fn(&FileInput{FilePath: input})
	elapsed = time.Since(start)
	return
}

type Day struct {
	PartOne func(Input) interface{}
	PartTwo func(Input) interface{}
}

func (d Day) Run() {
	input, part := GetFlags()
	_, f, _, _ := runtime.Caller(1)

	if !filepath.IsAbs(input) {
		input = filepath.Join(filepath.Dir(f), input)
	}
	var fn func(Input) interface{}
	if part == "one" {
		fn = d.PartOne
	} else if part == "two" {
		fn = d.PartTwo
	}
	result, elapsed := Execute(fn, input)
	fmt.Printf("part %s: %v (took: %v)\n", part, result, elapsed)

}

File: utils/execute.py
import time
from argparse import ArgumentParser
from collections.abc import Callable

from utils.input import Input


class Day:
    def __init__(
        self,
        part_one: Callable[[Input], int | None],
        part_two: Callable[[Input], int | None],
    ) -> None:
        self.part_one = part_one
        self.part_two = part_two

    def run(self) -> None:
        parser = ArgumentParser()
        parser.add_argument("-i", "--input", type=str)
        parser.add_argument("-p", "--part", type=str)
        parser.add_argument("-v", "--verbose", action="store_true")
        args = parser.parse_args()

        if args.part not in ("one", "two"):
            ValueError(f"invalid part {args.part}")

        fn = self.part_one if args.part == "one" else self.part_two

        start = time.perf_counter()
        result = fn(Input(args.input))
        elapsed = time.perf_counter() - start
        print(f"Part {args.part}: {result} (took: {elapsed:.2f}s)")
